lib/resumption/client.go

```diff
@@ -0,0 +1,362 @@
+
+// WrapSSHClientConn tries to detect if the server at the other end of nc is a
+// resumption v1 server, and if so it returns a [net.Conn] that will
+// transparently resume itself (using the provided redial func). If the
+// connection is wrapped, the context applies to the lifetime of the returned
+// connection, not just the duration of the function call.
+func WrapSSHClientConn(ctx context.Context, nc net.Conn, redial redialFunc) (net.Conn, error) {
+	dhKey, err := ecdh.P256().GenerateKey(rand.Reader)
+	if err != nil {
+		logrus.WithError(err).Error("Failed to generate ECDH key, proceeding without resumption (this is a bug).")
+		return nc, nil
+	}
+
+	// adds a read buffer with the ability to peek to nc
+	conn := ensureMultiplexerConn(nc)
+
+	// We must send the first 8 bytes of the version string to go through some
+	// older versions of the multiplexer that sits in front of the Teleport SSH
+	// server; thankfully, no matter which SSH client we'll end up using, it
+	// must send `SSH-2.0-` as its first 8 bytes, as per RFC 4253 ("The Secure
+	// Shell (SSH) Transport Layer Protocol") section 4.2. Sending only the
+	// first 8 bytes rather than a full version string is noncompliant behavior,
+	// but our SSH client is only intended to talk to Teleport-implemented SSH
+	// servers anyway, and other clients in the ecosystem do much worse
+	// (ssh-keyscan will wait for the server to send data before sending
+	// anything, for example).
+	if _, err := conn.Write([]byte(sshPrefix)); err != nil {
+		conn.Close()
+		return nil, trace.Wrap(err)
+	}
+
+	dhPub, hostID, err := readServerVersionExchange(conn)
+	if err != nil {
+		conn.Close()
+		return nil, trace.Wrap(err)
+	}
+
+	if dhPub == nil {
+		// regular SSH connection, conn is about to read the SSH- line from the
+		// server but we've sent sshPrefix already, so we have to skip it from
+		// the application side writes
+		logrus.Debug("Server does not support resumption, proceeding without.")
+		return &sshVersionSkipConn{
+			Conn:           conn,
+			alreadyWritten: sshPrefix,
+		}, nil
+	}
+
+	dhSecret, err := dhKey.ECDH(dhPub)
+	if err != nil {
+		logrus.WithError(err).Warn("Failed to complete ECDH key exchange, proceeding without resumption.")
+		return &sshVersionSkipConn{
+			Conn:           conn,
+			alreadyWritten: sshPrefix,
+		}, nil
+	}
+
+	otp32 := sha256.Sum256(dhSecret)
+	token := resumptionToken(otp32[:16])
+
+	if _, err := conn.Write([]byte(clientSuffixV1)); err != nil {
+		conn.Close()
+		return nil, trace.Wrap(err)
+	}
+	if _, err := conn.Write(dhKey.PublicKey().Bytes()); err != nil {
+		conn.Close()
+		return nil, trace.Wrap(err)
+	}
+	if _, err := conn.Write([]byte{newConnClientExchangeTag}); err != nil {
+		conn.Close()
+		return nil, trace.Wrap(err)
+	}
+
+	resumableConn := newResumableConn(conn.LocalAddr(), conn.RemoteAddr())
+	// runClientResumable expects a brand new, locked *Conn
+	resumableConn.mu.Lock()
```

Comment:
Is it really necessary for `runResumeV1Unlocking` to require a locked connection as input?  I feel like that design choice makes the entire locking model in this file unnecessarily confusing and difficult to trace.

Reply:
The first connection needs to attach before anything else can use the `Conn`, but we don't have a sensible way to synchronously wait for the underlying conn to be attached before returning the `Conn` or storing it somewhere where it's possible that something might attempt to attach to it again.

We're about 50/50 in whether or not the actual protocol handling should happen in a background goroutine or synchronous - and even when backgrounding, we often want control over what happens with the goroutine itself - so it's not super clear how to actually split the flow such that we can block until the first conn handler is marked as attached.

Besides, it's not like a single `r.mu.Lock()` at the beginning of the function changes things much.